src/Features/LanguageServer/ProtocolUnitTests/SemanticTokens/SemanticTokensEditsTests.cs

```diff
@@ -0,0 +1,204 @@
+        [Fact]
+        public async Task TestGetSemanticTokensEdits_ReturnMinimalEdits()
+        {
+            /*
+             * Original markup:
+             *     // Comment  
+             */
+            var updatedMarkup =
+@"{|caret:|}class
+// Comment";
+
+            using var workspace = CreateTestWorkspace(updatedMarkup, out var locations);
+            var results = await GetActualEdits(workspace, locations["caret"].First(), s_singleLineDataCase, previousResultId: "0");
+
+            // First edit: Replaces length of token (10 to 5) and replaces token type (comment to keyword)
+            var expectedFirstEdit = SemanticTokensEditsHandler.GenerateEdit(
+                start: 2, deleteCount: 2, data: new int[] { 5, SemanticTokensHelpers.GetTokenTypeIndex(LSP.SemanticTokenTypes.Keyword) });
+
+            // Second edit: Creates new token for '// Comment'
+            var expectedSecondEdit = SemanticTokensEditsHandler.GenerateEdit(
+                start: 5, deleteCount: 0, data: new int[] { 1, 0, 10, SemanticTokensHelpers.GetTokenTypeIndex(LSP.SemanticTokenTypes.Comment), 0 });
+
+            Assert.Equal(expectedFirstEdit, ((LSP.SemanticTokensEdits)results).Edits[0]);
+            Assert.Equal(expectedSecondEdit, ((LSP.SemanticTokensEdits)results).Edits[1]);
+            Assert.Equal("1", ((LSP.SemanticTokensEdits)results).ResultId);
+        }
+
+        /// <summary>
+        /// Tests to make sure that if we don't have a matching semantic token set for the document in the cache,
+        /// we return the full set of semantic tokens.
+        /// </summary>
+        [Fact]
+        public async Task TestGetSemanticTokensEditsNoCacheAsync()
+        {
+            /*
+             * Original markup:
+             *     // Comment
+             *     static class C { }  
+             */
+            var updatedMarkup =
+@"{|caret:|}// Comment
+
+static class C { }";
+
+            using var workspace = CreateTestWorkspace(updatedMarkup, out var locations);
+            var results = await GetActualEdits(workspace, locations["caret"].First(), s_basicDataCase, previousResultId: "10");
+
+            // Make sure we're returned SemanticTokens instead of SemanticTokensEdits.
+            Assert.True(results.Value is LSP.SemanticTokens);
+        }
+
+        private static async Task<SumType<LSP.SemanticTokens, LSP.SemanticTokensEdits>> GetActualEdits(
+            Workspace workspace,
+            LSP.Location caretLocation,
+            int[] originalData,
+            string previousResultId)
+        {
+            var cache = GetSemanticTokensCache(workspace);
+            var originalTokens = await GetOriginalTokens(caretLocation.Uri, originalData, cache);
+            await cache.UpdateCacheAsync(caretLocation.Uri, originalTokens, CancellationToken.None);
+
+            var results = await RunGetSemanticTokensEditsAsync(workspace.CurrentSolution, caretLocation, previousResultId);
+            return results;
+        }
```

Comment:
so, i don't like these tests ata ll :-/  we should chat about a better, but more involved, way of writing this stuff.